pulsar/producer.go

```diff
@@ -0,0 +1,167 @@
+	// Standard hashing functions available are:
+	//
+	//  - `JavaStringHash` : Java String.hashCode() equivalent
+	//  - `Murmur3_32Hash` : Use Murmur3 hashing function.
+	// 		https://en.wikipedia.org/wiki/MurmurHash">https://en.wikipedia.org/wiki/MurmurHash
+	//
+	// Default is `JavaStringHash`.
+	HashingScheme
+
+	// Set the compression type for the producer.
+	// By default, message payloads are not compressed. Supported compression types are:
+	//  - LZ4
+	//  - ZLIB
+	//  - ZSTD
+	//
+	// Note: ZSTD is supported since Pulsar 2.3. Consumers will need to be at least at that
+	// release in order to be able to receive messages compressed with ZSTD.
+	CompressionType
+
+	// Set a custom message routing policy by passing an implementation of MessageRouter
+	// The router is a function that given a particular message and the topic metadata, returns the
+	// partition index where the message should be routed to
+	MessageRouter func(Message, TopicMetadata) int
+
+	// Control whether automatic batching of messages is enabled for the producer. By default batching
+	// is enabled.
+	//
+	// When batching is enabled, multiple calls to Producer.sendAsync can result in a single batch to be sent to the
+	// broker, leading to better throughput, especially when publishing small messages. If compression is enabled,
+	// messages will be compressed at the batch level, leading to a much better compression ratio for similar headers or
+	// contents.
+	//
+	// When enabled default batch delay is set to 1 ms and default batch size is 1000 messages
+	// Setting `DisableBatching: true` will make the producer to send messages individually
+	DisableBatching bool
+
+	// Set the time period within which the messages sent will be batched (default: 10ms) if batch messages are
+	// enabled. If set to a non zero value, messages will be queued until this time interval or until
+	BatchingMaxPublishDelay time.Duration
+
+	// Set the maximum number of messages permitted in a batch. (default: 1000) If set to a value greater than 1,
+	// messages will be queued until this threshold is reached or batch interval has elapsed
+	BatchingMaxMessages uint
+}
+
+// The producer is used to publish messages on a topic
+type Producer interface {
+	// return the topic to which producer is publishing to
+	Topic() string
+
+	// return the producer name which could have been assigned by the system or specified by the client
+	Name() string
+
+	// Send a message
+	// This call will be blocking until is successfully acknowledged by the Pulsar broker.
+	// Example:
+	// producer.Send(ctx, pulsar.ProducerMessage{ Payload: myPayload })
+	Send(context.Context, *ProducerMessage) error
+
+	// Send a message in asynchronous mode
+	// The callback will report back the message being published and
+	// the eventual error in publishing
+	SendAsync(context.Context, *ProducerMessage, func(MessageID, *ProducerMessage, error))
```

Comment:
Here we can see that the interface exposed to the user is consistent with the cgo version of pulsar-client-go. I can understand that the benefit of this is compatible with the cgo version of pulsar-client-go, so that when the user is replacing Just replace the lib you depend on. But the operation of async, we use channel to deal with it is more reasonable? Similar like this:

```
Type XXXAsync interface {
        Input() chan<- *ProducerMessage
        Successes() <-chan *ProducerMessage
}
```

Reply:
Regarding this question, I am thinking, we are strictly following the cgo version of the API interface, or try to keep the interface with it, and more to write code according to the style of the go community.

Reply:
IMO it is better to design an API that is more native to go developers.

Reply:
Sure, the current code just imported the existing API and plugged the new implementation on that. 

 Since it’s a new repo, we don’t need to keep company with the existing go library. And until this gets officially released, we can keep iterating over the API without worries.